## Lab: SQL Practice

> Authors: Matt Brems & Jeff Hale

In this lab, you have 11 prompts. Most prompt will require you to do two things:
1. Write a SQL query.
2. Use the result of that query to answer the question.

The purpose of this lab is to help you practice your SQL - an **important** skill for working with data! 

Do everything with an SQL query in `pd.read_sql()`. 

If you want extra practice, do everything in pandas, too! 🐼

**Data**: The data used in this lab is the Iowa liquor database. It contains various tables of transactions. The ERD for this database is here:

<img src="./images/erd.png" alt="erd" width="750"/>

In [1]:
# Imports

import pandas as pd
from sqlalchemy import create_engine

In [2]:
# Set up the engine to access the data.

# 'driver://username:password@host:port/database'
engine = create_engine('postgres://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are there in the `stores` table of the Iowa liquor database?

In [3]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT COUNT(*)
FROM stores
"""

pd.read_sql(sql, engine)

count
0   1973

**Prompt 1 Answer**: There are 1973 rows

### Prompt 2

If you sort the stores in alphabetical order by name, which store is fourth? It's okay that numbers and capital letters come before lower case letter.

In [4]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.name
FROM stores AS s
ORDER BY s.name ASC
LIMIT 4
"""

pd.read_sql(sql, engine)

name
0  218 Fuel Express & Chubby's Liquor
1          3047 Cub Foods / Iowa City
2               3060 Cub Foods / Ames
3         3061 Cub Foods / Sioux City

**Prompt 2 Answer**: 3061 Cub Foods / Sioux City

### Prompt 3

For sales where the category name is `IMPORTED VODKA`, if you sort by vendor in alphabetical order, what are the bottle price and number of bottles in each of the first 3 transactions?

In [5]:
sql = """
SELECT s.category_name, s.vendor, s.btl_price, s.bottle_qty
FROM sales AS s
WHERE s.category_name ILIKE "IMPORTED VODKA"
LIMIT 3;
"""

In [6]:
sql = """
SELECT s.*
FROM sales AS s
LIMIT 10;
"""

In [7]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.category_name, s.vendor, s.btl_price, s.bottle_qty
FROM sales AS s
WHERE s.category_name ILIKE 'imported vodka'
ORDER BY vendor
LIMIT 4;
"""

pd.read_sql(sql, engine)

category_name                     vendor btl_price  bottle_qty
0  IMPORTED VODKA  Adamba Imports Int'l Inc.     $8.99          12
1  IMPORTED VODKA      A Hardy / U.S.A. Ltd.    $28.50           3
2  IMPORTED VODKA      A Hardy / U.S.A. Ltd.    $28.50           3
3  IMPORTED VODKA      A Hardy / U.S.A. Ltd.    $28.50           3

**Prompt 3 Answer**: (8.99, 12), (28.50, 3), (28.50, 1)

### Prompt 4

What is the total amount of liquor sold in February 2015? Display it in dollars and cents.

In [8]:
sql = """
SELECT FORMAT(SUM(s.total)) AS "Total"
FROM sales AS s
WHERE s.date > '2015-02-01' AND s.date < '2015-02-28'
LIMIT 10;
"""


In [44]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT CAST(SUM(s.total) AS money)
FROM sales AS s
WHERE s.date > '2015-02-01' AND s.date < '2015-02-28'
LIMIT 10;
"""

pd.read_sql(sql, engine)

sum
0  $21,295,350.55

**Prompt 4 Answer**: The total amount of liquor sold in February 2015 is '$21295350.55'

### Prompt 5

Among all transactions where `IMPORTED VODKA` was sold, which vendor has the most transactions? 

Report the total number of transactions, the number of bottles sold, and the total amount of revenue from these transactions. 

Rename the columns to make sure there isn't any confusion as to what they mean.

In [10]:
sql = """
SELECT s.vendor, s.bottle_qty, s.total
FROM sales AS s
WHERE s.category_name ILIKE 'imported vodka'
GROUP BY COUNT(s.vendor)
LIMIT 10
"""

In [11]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.vendor, COUNT(s.vendor) AS "transactions", SUM(s.bottle_qty) AS "bottle_sold", SUM(s.total) AS "revenue"
FROM sales AS s
WHERE s.category_name ILIKE 'imported vodka'
GROUP BY s.vendor
ORDER BY COUNT(s.vendor) DESC
LIMIT 10
"""

pd.read_sql(sql, engine)

vendor  transactions  bottle_sold     revenue
0  Pernod Ricard USA/Austin Nichols         42338       403450  7880365.88
1               Bacardi U.S.A. Inc.         26441       236047  5771242.07
2                   Diageo Americas         15725       152038  3346425.42
3   Constellation Wine Company Inc.         12565       116053  1895005.51
4                   Jim Beam Brands          5850       104119  1807097.23
5                       Stoli Group          4443        35513   832071.19
6              Bacardi U.S.A., Inc.          3344        30239   765347.31
7              Imperial Brands Inc.          2799        21033   265317.79
8            Moet Hennessy USA Inc.          2153        10966   323126.54
9  Constellation Wine Company, Inc.          1672        12956   200283.64

**Prompt 5 Answer**: 

Most Transactions:
Pernod Ricard USA/Austin Nichols, 42338 transactions, 403450 bottles sold,$ 7880365.88

### Prompt 6

The reason we sell things is to make money. In sales, the term **markup** means the amount of money charged for a product over the cost to make that product.

In Iowa, the law states the minimum price at which liquor may be sold.

Create a column called `Markup` that shows the difference between the `btl_price` and the `state_btl_cost`.

In [12]:
# Use a SQL query to display the table discussed above.

sql = """
SELECT s.*, (s.btl_price - s.state_btl_cost) AS "markup"
FROM sales as s
LIMIT 4
"""

pd.read_sql(sql, engine)

date convenience_store  store county_number county category  \
0 2014-11-13              None   2635            82  Scott  1032200   
1 2014-11-13              None   2635            82  Scott  1032200   
2 2014-11-13              None   2635            82  Scott  1032200   
3 2014-11-13              None   2635            82  Scott  1032080   

           category_name vendor_no                            vendor   item  \
0  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34036   
1  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34076   
2  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34116   
3         IMPORTED VODKA       420            Moet Hennessy USA Inc.  34155   

       description  pack  liter_size state_btl_cost btl_price  bottle_qty  \
0    Absolut Pears    12         750         $11.49    $17.24           4   
1  Absolut Vanilia    12         750         $11.49    $17.24          12   
2  Absolut Mandrin    12         750         $11.49    $17.24          12   
3  Belvedere Vodka     6         750         $17.15    $25.73           6   

    total markup  
0   68.96  $5.75  
1  206.88  $5.75  
2  206.88  $5.75  
3  154.38  $8.58

### Prompt 7

Calculate and display the "Percentage Markup" per store (just store id, not name) by dividing markup by the state bottle cost. 

This will tell you how much stores are increasing their average price over the state minimum.

In [13]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.*, ((s.btl_price - s.state_btl_cost) / s.state_btl_cost) AS "percent markup"
FROM sales as s
LIMIT 4
"""

pd.read_sql(sql, engine)

date convenience_store  store county_number county category  \
0 2014-11-13              None   2635            82  Scott  1032200   
1 2014-11-13              None   2635            82  Scott  1032200   
2 2014-11-13              None   2635            82  Scott  1032200   
3 2014-11-13              None   2635            82  Scott  1032080   

           category_name vendor_no                            vendor   item  \
0  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34036   
1  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34076   
2  IMPORTED VODKA - MISC       370  Pernod Ricard USA/Austin Nichols  34116   
3         IMPORTED VODKA       420            Moet Hennessy USA Inc.  34155   

       description  pack  liter_size state_btl_cost btl_price  bottle_qty  \
0    Absolut Pears    12         750         $11.49    $17.24           4   
1  Absolut Vanilia    12         750         $11.49    $17.24          12   
2  Absolut Mandrin    12         750         $11.49    $17.24          12   
3  Belvedere Vodka     6         750         $17.15    $25.73           6   

    total  percent markup  
0   68.96        0.500435  
1  206.88        0.500435  
2  206.88        0.500435  
3  154.38        0.500292

### Prompt 8

Calculate the average percentage markup per **store**. Return the average percentage markup and store number for the three stores with the largest markups.

In [14]:
## claire helped me here

sql = """
SELECT s.store,
    stores.name,
    stores.store_address,
    AVG((s.btl_price - s.state_btl_cost) / s.state_btl_cost) AS "Percentage Markup",
    MIN((s.btl_price - s.state_btl_cost) / s.state_btl_cost) AS "Min Markup",
    MAX((s.btl_price - s.state_btl_cost) / s.state_btl_cost) AS "Max Markup"  
FROM sales AS s
LEFT JOIN stores
    ON s.store = stores.store
GROUP BY s.store, stores.name, stores.store_address
ORDER BY AVG((s.btl_price - s.state_btl_cost) / s.state_btl_cost) DESC
LIMIT 3
"""

# she said to use an average and it fixed by entire problem by dividng by zero

In [15]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.store, AVG((s.btl_price - s.state_btl_cost) / s.state_btl_cost) AS "markup"
FROM sales AS s
GROUP BY s.store
ORDER BY AVG((s.btl_price - s.state_btl_cost) / s.state_btl_cost) DESC
LIMIT 10
"""

pd.read_sql(sql, engine)

store    markup
0   4013  0.525471
1   4024  0.521499
2   4266  0.518903
3   3833  0.515094
4   3660  0.514591
5   3628  0.514439
6   3889  0.514126
7   3934  0.514098
8   3749  0.513747
9   3696  0.513286

### **Prompt 8 Answer**: The store with the highest average markup percentage is store 4013, 4024, 4266

### Prompt 9

Retrieve the three stores with the highest average markup percentage. Along with the store IDs and average markup percentage, return the minimum and maximum markup percentage for those stores, the names and addresses of those store.

What are the names of the 3 stores?

In [16]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.store, stores.name,
    stores.store_address, 
    AVG((s.btl_price - s.state_btl_cost) / s.state_btl_cost) AS "markup", 
    MIN((s.btl_price - s.state_btl_cost) / s.state_btl_cost) AS "min_markup",
    MAX((s.btl_price - s.state_btl_cost) / s.state_btl_cost) AS "max_markup" 
FROM sales AS s
LEFT JOIN stores
    ON s.store = stores.store
GROUP BY s.store, stores.store_address, stores.name
ORDER BY AVG((s.btl_price - s.state_btl_cost) / s.state_btl_cost) DESC
LIMIT 4
"""

pd.read_sql(sql, engine)

store                        name  \
0   4013      Wal-Mart 0841 / Tipton   
1   4024  Wal-Mart 1546 / Iowa Falls   
2   4266  Wal-Mart 1683 / Shenandoah   
3   3833    Wal-Mart 3394 / Atlantic   

                                       store_address    markup  min_markup  \
0  1126 Highway 38 North\nTipton, IA 527720000\n(...  0.525471    0.498571   
1  840 S Oak\nIowa Falls, IA 501260000\n(42.50295...  0.521499    0.498462   
2  705 S Fremont\nShenandoah, IA 516010000\n(40.7...  0.518903    0.498462   
3  1905 East 7th St\nAtlantic, IA 500220000\n(41....  0.515094    0.498462   

   max_markup  
0    6.708763  
1    6.708763  
2    6.708763  
3    6.708763

**Prompt 9 Answer**: The three stores with the highest markups are Walmart 0841, Walmart 1546, Walmart 1683

### Prompt 10

Which nine counties sell the highest number of bottles of liquor per capita? Calculate the average bottle size of the bottles of liquor sold by these counties, rounded to the nearest first decimal place. Also display the county name, and average bottle size.

Of the nine counties selling the highest number of bottles of liquor per capita, which county sells (on average) the largest bottle size?

In [17]:
sql = """
SELECT s.county, s.total, c.population
FROM sales AS s
LEFT JOIN counties as c
    ON s.county = c.county
GROUP BY s.county
LIMIT 10
"""

In [33]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT s.county, SUM(s.bottle_qty) / c.population AS "bottles per capitia", AVG(p.bottle_size) AS "average bottle size"
FROM sales AS s
LEFT JOIN counties as c
    ON s.county = c.county
LEFT JOIN products as p
    ON s.item = p.item_no
GROUP BY s.county, c.population
ORDER BY SUM(s.bottle_qty) / c.population DESC
LIMIT 10
"""

pd.read_sql(sql, engine)

county  bottles per capitia  average bottle size
0           None                  NaN           774.912740
1      Dickinson                 22.0           974.148890
2     Black Hawk                 15.0           862.902896
3           Polk                 15.0           892.588027
4    Cerro Gordo                 14.0           986.427392
5          Scott                 13.0           871.626636
6        Johnson                 13.0           895.373151
7           Linn                 12.0           886.254321
8        Kossuth                 11.0          1041.042443
9  Pottawattamie                 11.0           893.713467

**Prompt 10 Answer**: Dickinson, 22 Bottles per capitia @ 974 mL

### Prompt 11

In the `sales` table, three columns are `btl_price`, `bottle_qty`, and `total`. The `total` for a transaction _should be_ the product of `btl_price` and `bottle_qty`. How many transactions, if any, have a value of `total` that is not equal to `btl_price` times `bottle_qty`?

> You might get a type error. We cannot compare type `money` to type `real`. You may have to convert `total` to type `money` to compare the two directly.

In [ ]:
sql = """
SELECT 
    s.btl_price, 
    s.bottle_qty, 
    s.btl_price * s.bottle_qty AS "calculated total", 
    CAST(s.total AS money), 
    CASE
        WHEN s.btl_price * s.bottle_qty = CAST(s.total AS money) THEN 1
        ELSE 0
    END AS "equivalent"
FROM sales AS s
LIMIT 10
"""

In [68]:
# Use a SQL query to find the answer to the above prompt.

sql = """
SELECT SUM(
    CASE
        WHEN s.btl_price * s.bottle_qty = CAST(s.total AS money) THEN 1
        ELSE 0
    END),
    COUNT(s.date)
FROM sales AS s
LIMIT 10
"""


pd.read_sql(sql, engine)

sum    count
0  3049913  3049913

**Prompt 11 Answer**: According to my findings, all 3,049,913 btl_price, bottle_qty, and total are accurate